staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/defaulting/validation.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package defaulting
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/go-openapi/validate"
+
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	schemaobjectmeta "k8s.io/apiextensions-apiserver/pkg/apiserver/schema/objectmeta"
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/schema/pruning"
+	apiservervalidation "k8s.io/apiextensions-apiserver/pkg/apiserver/validation"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// ValidateDefaults checks that default values validate and are properly pruned.
+func ValidateDefaults(pth *field.Path, s *structuralschema.Structural) (field.ErrorList, error) {
+	f := NewRootObjectFunc().WithTypeMeta(metav1.TypeMeta{APIVersion: "validation/v1", Kind: "Validation"})
+	v := validator{s}
+	return v.validate(pth, s, f, false)
+}
+
+type validator struct {
+	rootSchema *structuralschema.Structural
+}
+
+// validate is the recursive step func for the validation. insideMeta is true if s specifies TypeMeta or ObjectMeta.
+func (v *validator) validate(pth *field.Path, s *structuralschema.Structural, f SurroundingObjectFunc, insideMeta bool) (field.ErrorList, error) {
+	if s == nil {
+		return nil, nil
+	}
+
+	if s.XEmbeddedResource {
+		insideMeta = false
+		f = NewRootObjectFunc().WithTypeMeta(metav1.TypeMeta{APIVersion: "validation/v1", Kind: "Validation"})
+		v2 := *v
+		v = &v2
+		v.rootSchema = s
+	}
+
+	allErrs := field.ErrorList{}
+
+	if s.Default.Object != nil {
+		validator := validate.NewSchemaValidator(s.ToGoOpenAPI(), nil, "", strfmt.Default)
+
+		if insideMeta {
+			obj, _, err := f(runtime.DeepCopyJSONValue(s.Default.Object))
+			if err != nil {
+				return nil, fmt.Errorf("failed to prune default value: %v", err)
```

Comment:
nit: the step here doesn't look like pruning the default value

this is composing a test object with proper typemeta field for the later objectmeta validation

Reply:
it's one of these "should never happen" cases.

Reply:
added comment why